app/DynamicGraphs/GraphGenerator.hs

```diff
@@ -88,25 +101,31 @@ reqToStmts' parentID (RAW rawText) = do
 reqToStmts' _ (FCES _ _) = return []
 
 
-makeNode :: Text -> State Integer (DotNode Text)
+makeNode :: Text -> State GeneratorState (DotNode Text)
 makeNode name = do
-    i <- State.get
-    _ <- State.put (i + 1)
-    return $ DotNode (mappendTextWithCounter name i) [AC.Label $ toLabelValue name]
-
-makeBool :: Text -> State Integer (DotNode Text)
+    GeneratorState i nodesMap <- State.get
+    case Map.lookup name nodesMap of
+        Nothing -> do
+            let node = DotNode
+                        (mappendTextWithCounter name i)
+                        [AC.Label $ toLabelValue name]
+                nodesMap' = Map.insert name node nodesMap
+            State.put (GeneratorState (i + 1) nodesMap')
+            return node
+        Just node -> return node
+
+makeBool :: Text -> State GeneratorState (DotNode Text)
 makeBool text1 = do
-    i <- State.get
-    _ <- State.put (i + 1)
+    GeneratorState i nodesMap <- State.get
```

Comment:
You can use `State.modify` here as well.

Reply:
`i` actually gets used below on line 121, so I believe I do need to explicitly access the state before updating it.